content/blog/2019-10-23-loop-unrolling-with-bril.md

```diff
@@ -0,0 +1,154 @@
+
+### Implementation details
+
+#### External Packages
+In this project, we use two external packages. The first one is the array extension of Bril, whose source code can be found [here](https://github.com/Checkmate50/bril/tree/arrays). The second one is the C code generation for Bril. The source code can be found [here](https://github.com/xu3kev/bril2c).
+
+#### Static Analysis
+This section details how we implement the static loop unrolling pass. We first state our definition of "regular loops" that
+can be unrolled by our pass. A "regular loop" should have the following properties other than the ones mentioned in the 
+input specification section:
+- The control flow can only leave the loop from either of the two blocks connected by the back-edge. 
+- The loop can not contain any sub-loop. 
+
+In this section, without special mentioning, we will use *entry* to denote the entrance block of the loop (destination of 
+the backedge), and use *exit* to denote the exit block of the loop (source of the backedge). The implementation details of 
+each step outlined in the Methodology section are show below. 
+
+##### CFG Analysis
+We use the functions provided in ``examples/cfg.py`` to construct the CFG and compute the predecessors and successors of 
+each basic block. The ``block_map`` function is slightly modified to give each basic block a label, inserted as the first
+instruction of the block. We use the pass provided in ``examples/dom.py`` to compute the dominators of each basic block. 
+
+##### Dataflow Analysis
+We implemented a reaching definition pass in ``examples/df.py``, which is useful in finding the last instruction in the loop
+that updates the induction variable and the condition variable. Our implementation conforms with the pseudo-code in the 
+lecture notes. We also augmented the constant propagation pass so that it can propagate through ``id`` and arithmetic 
+instructions. The results of the constant propagation pass is used to find the initial value of the induction variable, 
+the stride of the induction variable, and the loop bound. 
+
+##### Finding Loops
+Our pass detects loops by first finding a backedge and then adding nodes into the loop. A back-edge is defined as a directed
+edge pointing from basic block A to basic block B where B dominates A. After locating the back-edge, we add all nodes that
+can reach the *exit* and are dominated by *entry* into the loop. We also filter out loops that are not regular by our standard and loops with sub-loops in this step. 
+
+##### Computing Tripcount
```

Comment:
As above, try to stay consistent about whether this is "tripcount," "trip count," or "trip-count."